txnsync/txngroups.go

```diff
@@ -17,33 +17,202 @@
+	BitmaskRekeyTo bitmask          `codec:"rekeybm,allocbound=maxBitmaskSize"`
+}
+
+type encodedKeyregTxnFields struct {
+	_struct                 struct{}                          `codec:",omitempty,omitemptyarray"`
+	VotePK                  []crypto.OneTimeSignatureVerifier `codec:"votekey,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVotePK           bitmask                           `codec:"votekeybm,allocbound=maxBitmaskSize"`
+	SelectionPK             []crypto.VRFVerifier              `codec:"selkey,allocbound=maxEncodedTransactionGroup"`
+	BitmaskSelectionPK      bitmask                           `codec:"selkeybm,allocbound=maxBitmaskSize"`
+	VoteFirst               []basics.Round                    `codec:"votefst,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVoteFirst        bitmask                           `codec:"votefstbm,allocbound=maxBitmaskSize"`
+	VoteLast                []basics.Round                    `codec:"votelst,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVoteLast         bitmask                           `codec:"votelstbm,allocbound=maxBitmaskSize"`
+	VoteKeyDilution         []uint64                          `codec:"votekd,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVoteKeyDilution  bitmask                           `codec:"votekdbm,allocbound=maxBitmaskSize"`
+	Nonparticipation        []bool                            `codec:"nonpart,allocbound=maxEncodedTransactionGroup"`
+	BitmaskNonparticipation bitmask                           `codec:"nonpartbm,allocbound=maxBitmaskSize"`
+}
+
+type encodedPaymentTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Receiver        []basics.Address    `codec:"rcv,allocbound=maxEncodedTransactionGroup"`
+	BitmaskReceiver bitmask             `codec:"rcvbm,allocbound=maxBitmaskSize"`
+	Amount          []basics.MicroAlgos `codec:"amt,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAmount   bitmask             `codec:"amtbm,allocbound=maxBitmaskSize"`
+
+	CloseRemainderTo        []basics.Address `codec:"close,allocbound=maxEncodedTransactionGroup"`
+	BitmaskCloseRemainderTo bitmask          `codec:"closebm,allocbound=maxBitmaskSize"`
+}
+
+type encodedAssetConfigTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedAssetTransferTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedAssetFreezeTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedApplicationCallTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedCompactCertTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type bitmask []byte
+
+// assumed to be in mode 0, sets bit at index to 1
+func (b bitmask) SetBit(index int) {
+	byteIndex := index/8 + 1
+	b[byteIndex] |= 1 << (index % 8)
+}
+
+func (b bitmask) EntryExists(index int, hint int) (bool, int) {
+	option := 0
+	if len(b) > 0 {
```

Comment:
if `len(b)` == 0, then we can exit right here. no need to further conditionals, right ?